### PR TITLE
Use name hash for model identifiers

### PR DESCRIPTION
**Pull request by tarrencev**
This is a backward incompatible change to migrate model identifiers to be name hash (`keccak(name)`) rather than the name literal. This will be necessary as we move to fully qualified model names and is necessary to support longer model names too. It aligns model identifiers with event identifiers which also uses `keccak(event_name)`

Ref DOJ-226, #1629

### CODE REVIEW

File: crates/dojo-core/src/world.cairo [hunk 1]
```diff
@@ -450,16 +460,16 @@ mod world {
...
+            contract.serialize(ref keys);
             emit_event_syscall(keys.span(), values).unwrap_syscall();
         }
 
         /// Sets the model value for an entity.
         ///
         /// # Arguments
         ///
-        /// * `model` - The name of the model to be set.
+        /// * `model` - The name hash of the model to be set.
```
Comment (glihm) [inline crates/dojo-core/src/world.cairo:472]:
```suggestion
        /// * `model` - The selector of the model to be set.
```

Comment (glihm) [reply crates/dojo-core/src/world.cairo:472]: We could apply the selector terminology here, or name hash is voluntarily kept?

**Review by glihm — COMMENTED** [review]
Interesting change about the entity, having the possibility of retrieving the entity from the model contract itself, but pointing to a different world opens new things.

File: crates/dojo-core/src/model.cairo [hunk 2]
```diff
@@ -1,7 +1,11 @@
+use dojo::world::IWorldDispatcher;
 use starknet::SyscallResult;
 
 trait Model<T> {
-    fn name(self: @T) -> felt252;
+    fn entity(world: IWorldDispatcher, keys: Span<felt252>, layout: Span<u8>) -> T;
+    fn name(self: @T) -> ByteArray;
+    fn version(self: @T) -> u8;
```
[the remark this answers is not in the document]
Comment (tarrencev) [reply crates/dojo-core/src/model.cairo:7]: the developer shouldn't care about it ideally. if we can include it in the manifest then we can optionally generate code for backward compat

File: crates/dojo-lang/src/plugin.rs [hunk 3]
```diff
@@ -441,6 +430,7 @@ impl MacroPlugin for BuiltinDojoPlugin {
             "dojo::contract".to_string(),
             "key".to_string(),
             "computed".to_string(),
+            "dojo::model".to_string(),
```
[the remark this answers is not in the document]
Comment (tarrencev) [reply crates/dojo-lang/src/plugin.rs:433]: hmm that is a good point. if i don't include this it does fail

File: crates/dojo-core/src/world.cairo [hunk 4]
```diff
@@ -7,31 +7,25 @@ use dojo::resource_metadata::ResourceMetadata;
 trait IWorld<T> {
     fn metadata(self: @T, resource_id: felt252) -> ResourceMetadata;
     fn set_metadata(ref self: T, metadata: ResourceMetadata);
-    fn model(self: @T, name: felt252) -> (ClassHash, ContractAddress);
+    fn model(self: @T, selector: felt252) -> (ClassHash, ContractAddress);
```
[the remark this answers is not in the document]
Comment (tarrencev) [reply crates/dojo-core/src/world.cairo:10]: That seems like an improvement but let's save it for a follow up PR

File: crates/dojo-core/src/model.cairo (hunk 2 above)
[the remark this answers is not in the document]
Comment (glihm) [reply crates/dojo-core/src/model.cairo:7]: In this case, as it's only a view function without storage impact, I suggest we shouldn't limit ourselves to an u8 and use a felt252 instead.

**Comment by glihm** [comment]
Will be merged with #1970 on storage layout branch.

**Pull request closed without merging**